src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs

```diff
@@ -0,0 +1,521 @@
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        var s1 = (ReadOnlySpan<char>)"""";
+        var s2 = (ReadOnlySpan<char>)"""";
+
+        Console.Write(s1.Length == s2.Length);
+
+        s1 = (ReadOnlySpan<char>)(string)null;
+        s2 = (ReadOnlySpan<char>)(string)null;
+
+        Console.Write(s1.Length == s2.Length);
+    }
+}
+
+", TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "TrueTrue", verify: Verification.Passes).VerifyIL("Test.Main", @"
+{
+  // Code size       79 (0x4f)
+  .maxstack  2
+  .locals init (System.ReadOnlySpan<char> V_0, //s1
+                System.ReadOnlySpan<char> V_1) //s2
+  IL_0000:  ldstr      """"
+  IL_0005:  call       ""System.ReadOnlySpan<char> System.ReadOnlySpan<char>.op_Implicit(string)""
+  IL_000a:  stloc.0
+  IL_000b:  ldstr      """"
+  IL_0010:  call       ""System.ReadOnlySpan<char> System.ReadOnlySpan<char>.op_Implicit(string)""
+  IL_0015:  stloc.1
+  IL_0016:  ldloca.s   V_0
+  IL_0018:  call       ""int System.ReadOnlySpan<char>.Length.get""
+  IL_001d:  ldloca.s   V_1
+  IL_001f:  call       ""int System.ReadOnlySpan<char>.Length.get""
+  IL_0024:  ceq
+  IL_0026:  call       ""void System.Console.Write(bool)""
+  IL_002b:  ldnull
+  IL_002c:  call       ""System.ReadOnlySpan<char> System.ReadOnlySpan<char>.op_Implicit(string)""
+  IL_0031:  stloc.0
+  IL_0032:  ldnull
+  IL_0033:  call       ""System.ReadOnlySpan<char> System.ReadOnlySpan<char>.op_Implicit(string)""
+  IL_0038:  stloc.1
+  IL_0039:  ldloca.s   V_0
+  IL_003b:  call       ""int System.ReadOnlySpan<char>.Length.get""
+  IL_0040:  ldloca.s   V_1
+  IL_0042:  call       ""int System.ReadOnlySpan<char>.Length.get""
+  IL_0047:  ceq
+  IL_0049:  call       ""void System.Console.Write(bool)""
+  IL_004e:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(23358, "https://github.com/dotnet/roslyn/issues/23358")]
+        public void StringConstNotConstConv()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+
+class Test
+{
+    public static void Main()
+    {       
+        var s1 = (ReadOnlySpan<char>)Hello();
+        var s2 = (ReadOnlySpan<char>)""Hello"";
+
+        Console.Write(s1[1] == s2[1]);
+    }
+
+    private static string Hello() => ""Hello"";
+}
+
+", TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "True", verify: Verification.Fails).VerifyIL("Test.Main", @"
+{
+  // Code size       49 (0x31)
+  .maxstack  3
+  .locals init (System.ReadOnlySpan<char> V_0, //s1
+                System.ReadOnlySpan<char> V_1) //s2
+  IL_0000:  call       ""string Test.Hello()""
+  IL_0005:  call       ""System.ReadOnlySpan<char> System.ReadOnlySpan<char>.op_Implicit(string)""
+  IL_000a:  stloc.0
+  IL_000b:  ldsflda    ""<PrivateImplementationDetails>.__StaticArrayInitTypeSize=10 <PrivateImplementationDetails>.D2EFCBBA102ED3339947E85F4141EB08926E40E9""
```

Comment:
Do the PrivateImplementationDetails types have the right alignment in the final binary? (E.g. if the Span element type long, they should be aligned at 8 bytes).

Are there enough breadcrumbs for IL rewriting tools to maintain the alignment?

Reply:
We use the same for ordinary arrays inits.
Would that imply the use for span is ok or span has extra requirements?

Reply:
InitializeArray gets the breadcrumbs about the element size via the RuntimeFieldHandle argument.

And the old CoreCLR big endian path through InitializeArray (currently unused) is explicitly handling the misasligned case: https://github.com/dotnet/coreclr/blob/46ab1d132c9ad471d79afa20c188c2f9c85e5f20/src/classlibnative/bcltype/arraynative.cpp#L1442 .

The optimization is definitely fine for span of bytes - it would be very useful even for just that. But it may need more plumbing to work well for spans of larger types.

Reply:
Can this memory be written to?
I can align and re-end in a static ctor

Reply:
It cannot be written to. The static data are R/O in IL-only images.

Reply:
> Just need to verify if the builder itself is emitted on the right boundary and whether it is sufficient to result in aligned RVAs.

What are the IL rewriters going to do?

> copy at runtime to an array

I think that the runtime would need to deal with the endianess fixup, similar to how it is done for ArrayInitialize today. I do not think you can deal with it in a reasonable way on your own.

Reply:
Ok, assuming that runtime can fixup endiannes, it would need to know the size of the "element" in the blob. It cannot know it now.

We refer to the blobs via FieldRVA static fields. The fields have types of appropriate size - `byte`, `short`, `int`, `long` or a synthetic struct with `.size`.  
The types have no meaning in existing cases, only size matters. I.E.  4 shorts will be mapped via a `long` field...


What if for the blobs where we care about alignment/endiannes, we always use syntetic struct with `.size` specifying the size and `.pack` specifying the element size?

Would that be sufficient?


Alternatively the synthetic struct could have a dummy primitive field of appropriate size. - `byte`, ..., `long`.  
Or even the actual element type field, may be even N fields, although that seems a bit of an overkill, since there are often thousands of elements in these blobs.

> What are the IL rewriters going to do?

IL rewriters should align on 8 bytes and keep the rest as-is. Most likely they do that already. 
FWIW `System.Reflection.Metadata` exposes the desired alignment as `ManagedPEBuilder.MappedFieldDataAlignment`.

Reply:
> Alternatively the synthetic struct could have a dummy primitive field of appropriate size.

I think this would be my pick. There is prior art in managed C++. Managed C++ emits dummy fields like that to control alignment of opaque structs. `.pack` does not make any guarantees about alignment today.

> IL rewriters should align on 8 bytes and keep the rest as-is. Most likely they do that already.

It would be useful to confirm.

Should the availability of this optimization be controlled via new `RuntimeFeature` ?

Reply:
> Alternatively the synthetic struct could have a dummy primitive field of appropriate size.

Will add this to the PR, then. And add alignment tests.

Will have to check the rewriters, but I think that is not blocking, just need to tell them. 

Need to think about `RuntimeFeature`. Maybe.

Reply:
@alrz - FYI. Re: https://github.com/dotnet/roslyn/pull/24249
Whatever is decided here will have effect on the stackalloc initializers. The `cpblk` path in initializers is exposed to the endianness in the same way. 

Nothing here is actionable to your current PR right now, so this is mostly FYI, but whatever approach is used for big types will have to be used in stackalloc initializers as well.

Reply:
since we are constrained to 1-byte elements, this is irrelevant now

---
In reply to: [165826041](https://github.com/dotnet/roslyn/pull/24621#discussion_r165826041) [](ancestors = 165826041)